### DISCUSSION TITLE
Type hints and optimized mode

### DISCUSSION THREAD

**Discussion by naglis** [python/typing#1598] [Q&A]
While reviewing [a PR](https://github.com/geigi/cozy/pull/848/files#r1460270387) I came across a somewhat interesting case - the validity of the return type depending on whether Python is running in optimized mode (starting Python with `-O` or `PYTHONOPTIMIZE` environment variable being set). E.g.:
```python
import typing


def foo() -> typing.NoReturn:
    assert False


print(foo())
```

the return type of `foo` is not valid in optimized mode since the `assert` statement will be removed and calling `foo` would return `None`.

Checking the snippet with `mypy` (version 1.8.0) does not report/warn about the issue.

I have searched in the [documentation](https://docs.python.org/3/library/typing.html), PEPs [483](https://peps.python.org/pep-0483/) and [484](https://peps.python.org/pep-0484/), in this repository and on [`mypy`](https://github.com/python/mypy), but could not find optimized mode being mentioned. How should such a case be handled? Should the type hint of the return type of `foo` be somehow dynamic depending on [`__debug__`](https://docs.python.org/3/library/constants.html#debug__)? Or are type hints always considered in non-optimized mode?

**Comment by sobolevn** [comment]
Just use `raise AssertionError` instead of `assert`

**Comment by Daverball** [comment]
This is something a linter should warn you about and in fact [flake8-bugbear](https://github.com/PyCQA/flake8-bugbear) does exactly that with B011. I don't think it's reasonable to expect a type checker to worry about what may or may not change its behavior in optimized builds.
